### ISSUE TITLE
Interpret Unknown as object

### ISSUE THREAD

**Issue by thesadru** [microsoft/pyright#3650]
Pyright's `reportUnknown-` settings are very helpful but force the user to have redundant casts.

```py
import typing

def func(x: typing.Sequence[object]) -> None:
    ...

x: object = ...

if isinstance(x, typing.Sequence):
    reveal_type(x)  # Type of "x" is "Sequence[Unknown]"
    func(x)  # Unknown argument type shouldn't matter here
```

Current solution:
```py
if isinstance(x, typing.Sequence):
    x = typing.cast("typing.Sequence[object]", x)
    func(x)
```

**Comment by erictraut** [comment]
It's unfortunate that `isinstance` doesn't allow a specialized type to be passed as the second argument, like `Sequence[object]`.

Let me make sure I understand what solution you're proposing here. Are you suggesting that for `isinstance` type narrowing where there is a downcast to a generic type where the type arguments are unknown, pyright should substitute `object` for all type arguments rather than `Unknown`?

I think that solution is problematic for a couple of reasons:
1. It would create problems for types that have invariant type parameters. It works OK for `Sequence`, but is problematic for `list` because `list[object]` cannot be used in many cases where `list[Unknown]` (which is the equivalent of `list[Any]`) can.
2. It would create a hard-to-explain inconsistency in the way pyright handles unknown type arguments. In all other cases, it generates an `Unknown` for cases like this. To create an exception like this, we'd need to come up with a defensible principle for when such an exception would be applied, and I'm struggling to come up with such a principle here.

I'll note that the `reportUnknown-` diagnostic rules are intended for strict-mode type checking. As such, they are targeted at users who are fully bought in to the benefits of static type checking and are willing to make (smallish) changes to their code to guarantee the highest levels of type safety. So, while I don't like the need for `cast` calls, it is somewhat justified here.

While I empathize with this problem, I don't like the proposed solution. Let me know if you have any thoughts on my rationale above.

**Comment by thesadru** [comment]
I believe that in that case there should be a separate configuration for arguments that is more lenient. Allowing `Unknown` to be used but only be matched to `object` or `Any`. 

If we go there, new types of warnings should be introduced to make the difference between the options clearer.

**Comment by erictraut** [comment]
You're suggesting that when an expression used within a call argument consists of a lone variable, that variable should fall under some different type checking rule than it would if that same variable were used in any other context? That would be hard to justify or explain.

As I said, the `reportUnknown-` diagnostic checks are strict-mode checks. You can choose to leave them off if you don't think they provide sufficient value.

**Comment by thesadru** [comment]
The problem is that `Unknown` can be interpreted as anything with that setting turned off. While it is optional it's hard to go without it.

**Comment by erictraut** [comment]
Interpreting `Unknown` as `object` will produce many false positives. I don't think that's the right answer.

The problem here is specific to `isinstance` type narrowing because there's no way for you to express the intended specialized type due to the runtime limitations of `isinstance`. 

If the `isinstance` narrowing logic were to substitute `object` for all unknown type arguments, it would result in many false positives in basic type checking scenarios. If `isinstance` narrowing logic substitutes `Unknown` for all unknown type arguments, false positives are eliminated unless/until you enable the `reportUnknown-` diagnostic checks. Those are strict-mode checks, so the people who use them tend to be more knowledgeable about type checking and more willing to make changes to their code to accommodate static type checking. So I think we've made the right tradeoff here.

If you find that you are frequently downcasting an `object` to a `Sequence[object]` within your code, you could write a small TypeGuard function that implements your suggested solution.

```python
def is_sequence(v: object) -> TypeGuard[Sequence[object]):
    return isinstance(v, Sequence)
```

**Comment by beauxq** [comment]
I hope that we're still looking for better solutions to this problem. I keep running into it.

It's gradually turning into:
Here's our `TypeGuard` for `Iterable`
Here's our `TypeGuard` for `Sequence`
Here's our `TypeGuard` for `Container`
and everything else in `collections.abc`
and all our custom generics...

and other people on the team that use mypy and not pyright wonder why we're maintaining all these useless TypeGuards (because mypy narrows it to `Any` even in strict mode, which is not a good solution - not type safe)

maybe a new unknown like `UnknownFromCovariantGenericTypeVarNarrowingThisNameIsTotallyNotTooLong`

that acts like `object` in strict mode, and acts like `Any` in other modes

The number of false positives from treating it like `object` in strict mode would be less than the number of errors for reporting it as a normal `Unknown`.
